977 pagination small

Martskin opened the pull request:
Closes #977

I took a stab at adding a secondary pagination component as navigation element per issue #977

https://cl.ly/f3dbe087c62b

#### Changelog

**New**

- Pagination Nav component

#### Testing / Reviewing

- [ ] Page interactive elements can be `<button>` or `<a>` tags.
- [ ] Active page changes when clicked
- [ ] Active page changes when < previous/next > elements are clicked.

Martskin committed:
feat(pagination): adds standalone/small pagination nav control

Martskin committed:
feat(pagination): adds pagination nav js

Martskin committed:
feat(pagination): adds js spec to doc site

Martskin committed:
feat(pagination): adds unit test for pagination nav

Martskin committed:
Merge branch 'master' into 977-pagination-small

# Conflicts:
#	src/globals/js/components.js
#	tests/axe/allHtml/a11y-html.json

Martskin committed:
Merge branch 'master' into 977-pagination-small

Martskin committed:
Merge branch 'master' into 977-pagination-small

Martskin committed:
Merge branch 'master' into 977-pagination-small

Martskin committed:
fix(codeexample): fixes showBtn state

asudoh commented on src/components/pagination-nav/pagination-nav.js:
```diff
@@ -0,0 +1,153 @@
+import settings from '../../globals/js/settings';
+import mixin from '../../globals/js/misc/mixin';
+import createComponent from '../../globals/js/mixins/create-component';
+import InitComponentBySearch from '../../globals/js/mixins/init-component-by-search';
```
The import here is a function instead of a class, thus we use camel case instead of pascal case.

asudoh commented on src/components/pagination-nav/pagination-nav.js:
```diff
@@ -0,0 +1,153 @@
+   */
+  constructor(element, options) {
+    super(element, options);
+    this.element.addEventListener('click', evt => this.handleClick(evt));
```
Carbon vanilla codebase has lifecycle management - `release()` method of this class should remove the event listener, to avoid event listener leak. We use [`on()`](https://github.com/IBM/carbon-components/blob/v9.70.0/src/globals/js/misc/on.js#L8-L16) utility function along with [`handles` mix-in](https://github.com/IBM/carbon-components/blob/v9.70.0/src/globals/js/mixins/handles.js) for that lifecycle management. Could you please do the same?

asudoh commented on src/components/pagination-nav/pagination-nav.js:
```diff
@@ -0,0 +1,153 @@
+    this.selectorPageDirectionArray = this.element.querySelectorAll(this.options.selectorPageDirection);
+    this.selectorPageSelect = this.element.querySelector(this.options.selectorPageSelect);
+    if (this.selectorPageSelect) {
+      this.selectorPageSelect.addEventListener('change', evt => this.handleSelectChange(evt));
```
Carbon vanilla codebase has lifecycle management - `release()` method of this class should remove the event listener, to avoid event listener leak. We use [`on()`](https://github.com/IBM/carbon-components/blob/v9.70.0/src/globals/js/misc/on.js#L8-L16) utility function along with [`handles` mix-in](https://github.com/IBM/carbon-components/blob/v9.70.0/src/globals/js/mixins/handles.js) for that lifecycle management. Could you please do the same?

asudoh commented on src/components/pagination-nav/pagination-nav.js:
```diff
@@ -0,0 +1,153 @@
+  constructor(element, options) {
+    super(element, options);
+    this.element.addEventListener('click', evt => this.handleClick(evt));
+    this.selectorPageElementArray = this.element.querySelectorAll(this.options.selectorPageElement);
```
I'd suggest not caching elements `[data-page]` here, considering dynamic DOM updates.

asudoh commented on src/components/pagination-nav/pagination-nav.js:
```diff
@@ -0,0 +1,153 @@
+  handleClick = evt => {
+    this.selectorPageElementArray.forEach((element, index) => {
+      if (element.classList.contains(this.options.classActive)) {
+        this.currentIndex = index;
```
This logic seems to be trying to get the previous index, but does not seem to cope with the selection made in the select box? e.g. Choosing `8` in the select box and hitting left arrow causes moving to `10` instead of `7`.

Martskin replied:
Yeah I wasn't sure what the expected behavior there would be so with the previous next arrows when the select menu is present. I was treating it more like prototype behavior. I'll mature it with this in mind!

asudoh replied:
I see - Thank you for clarifying! I think we should update the selection in the select box.

asudoh commented on src/components/pagination-nav/pagination-nav.js:
```diff
@@ -0,0 +1,153 @@
+    this.element.addEventListener('click', evt => this.handleClick(evt));
+    this.selectorPageElementArray = this.element.querySelectorAll(this.options.selectorPageElement);
+    this.totalPages = this.selectorPageElementArray.length;
+    this.currentIndex = 0;
```
I'd suggest not maintaining current index as an instance property, considering dynamic DOM updates. Instead, I'd suggest creating a method to determine the current index.

asudoh reviewed (commented):
Thank you for jumping @Martskin and it's amazing to see you understand vanilla codebase very quickly and well! Left some comments. Don't hesitate to ping me if you have any questions - Thanks!

Martskin commented:
> Thank you for jumping @Martskin and it's amazing to see you understand vanilla codebase very quickly and well! Left some comments. Don't hesitate to ping me if you have any questions - Thanks!

It's my pleasure! Thanks for the feedback and kind words! @asudoh 
I'll do some refactoring and get these changes pushed up.

Martskin committed:
Merge branch 'master' into 977-pagination-small

Martskin committed:
feat(pagination-nav): refactors js for pagination nav per PR feedback

Martskin committed:
Merge branch 'master' into 977-pagination-small

Martskin commented:
I refactored per your feedback @asudoh . Thanks again!
https://deploy-preview-1812--the-carbon-components.netlify.com/?nav=pagination-nav

asudoh commented on src/components/pagination-nav/pagination-nav.js:
```diff
@@ -0,0 +1,200 @@
+    super(element, options);
+    this.manage(on(this.element, 'click', evt => this.handleClick(evt)));
+    this.activePageNumber = this.initActivePageNumber();
+    this.selectorPageSelect = this.element.querySelector(this.options.selectorPageSelect);
```
I suggest not caching `<select>` here as an instance variable, to cope with dynamic DOM change.

asudoh commented on src/components/pagination-nav/pagination-nav.js:
```diff
@@ -0,0 +1,200 @@
+    this.clearActivePage(evt);
+    const selectorPageElementArray = this.element.querySelectorAll(this.options.selectorPageElement);
+    if (evt.target.matches(this.options.selectorPageButton)) {
+      this.activePageNumber = Number(evt.target.dataset.page);
```
Instead of caching active page number, let's call `initActivePageNumber()` as needed.

asudoh commented on src/components/pagination-nav/pagination-nav.js:
```diff
@@ -0,0 +1,200 @@
+      if (evt.target.dataset.pageDirection === 'previous') {
+        this.activePageNumber -= 1;
+      }
+      if (evt.target.dataset.pageDirection === 'next') {
```
Given we make `[data-page-direction]` selector configurable, let's define `attribPageDirection` option.

asudoh commented on src/components/pagination-nav/pagination-nav.js:
```diff
@@ -0,0 +1,200 @@
+    let pageNum;
+    const activePageElement = this.element.querySelector(this.options.selectorPageActive);
+    if (activePageElement) {
+      pageNum = Number(activePageElement.dataset.page);
```
Given we make `[data-page]` selector configurable, let's define `attribPage` option.

asudoh commented on src/components/pagination-nav/pagination-nav.js:
```diff
@@ -0,0 +1,200 @@
+      }
+    }
+    const selectorPageTarget = selectorPageElementArray[this.activePageNumber - 1];
+    selectorPageTarget.dataset.active = true;
```
Given we make `[data-active]` attribute configurable (with `[data-active="true"]` selector), let's define `attribActive` option.

asudoh commented on src/components/pagination-nav/pagination-nav.js:
```diff
@@ -0,0 +1,200 @@
+    this.clearActivePage(evt);
+    this.activePageNumber = Number(evt.target.value);
+    const selectorPageSelectOptions = this.selectorPageSelect.options;
+    selectorPageSelectOptions[selectorPageSelectOptions.selectedIndex].dataset.active = true;
```
Given we make `[data-active]` attribute configurable (with `[data-active="true"]` selector), let's define `attribActive` option.

asudoh commented on src/components/pagination-nav/pagination-nav.js:
```diff
@@ -0,0 +1,200 @@
+    selectorPageButtonArray.forEach(el => {
+      el.classList.remove(this.options.classActive, this.options.classDisabled);
+      el.disabled = false;
+      el.removeAttribute('data-active');
```
Given we make `[data-active]` attribute configurable (with `[data-active="true"]` selector), let's define `attribActive` option.

asudoh commented on src/components/pagination-nav/pagination-nav.js:
```diff
@@ -0,0 +1,200 @@
+    if (this.selectorPageSelect) {
+      const selectorPageSelectOptions = this.selectorPageSelect.options;
+      for (let i = 0; i < selectorPageSelectOptions.length; i++) {
+        selectorPageSelectOptions[i].removeAttribute('data-active');
```
Given we make `[data-active]` attribute configurable (with `[data-active="true"]` selector), let's define `attribActive` option.

asudoh commented on src/components/pagination-nav/pagination-nav.js:
```diff
@@ -0,0 +1,200 @@
+  constructor(element, options) {
+    super(element, options);
+    this.manage(on(this.element, 'click', evt => this.handleClick(evt)));
+    this.activePageNumber = this.initActivePageNumber();
```
Instead of caching what `initActivePageNumber()` returns, let's call that function as needed.

asudoh commented on src/components/pagination-nav/pagination-nav.js:
```diff
@@ -0,0 +1,200 @@
+  /**
+   * Get active page number
+   */
+  initActivePageNumber = () => {
```
Naming nit - Probably `getActivePageNumber()`?

asudoh commented on src/components/pagination-nav/pagination-nav.js:
```diff
@@ -0,0 +1,200 @@
+   */
+  clearActivePage = evt => {
+    const selectorPageButtonArray = this.element.querySelectorAll(this.options.selectorPageButton);
+    selectorPageButtonArray.forEach(el => {
```
What `qSA` returns `NodeList` does not have `forEach()` in IE11 - We can use `Array.prototype.forEach.call()`.

asudoh commented on src/components/pagination-nav/pagination-nav.js:
```diff
@@ -0,0 +1,200 @@
+    });
+    if (this.selectorPageSelect) {
+      const selectorPageSelectOptions = this.selectorPageSelect.options;
+      for (let i = 0; i < selectorPageSelectOptions.length; i++) {
```
Nit - Seems that regular `for()` call and `forEach()` calls are both in use. Should we unify the style? Basically this codebase prefers the latter.

asudoh commented on src/components/pagination-nav/pagination-nav.js:
```diff
@@ -0,0 +1,200 @@
+    }
+    if (evt.target.matches(this.options.selectorPageDirection)) {
+      if (evt.target.dataset.pageDirection === 'previous') {
+        this.activePageNumber -= 1;
```
Instead of caching active page number, let's call `initActivePageNumber()` as needed.

asudoh commented on src/components/pagination-nav/pagination-nav.js:
```diff
@@ -0,0 +1,200 @@
+   */
+  handleClick = evt => {
+    this.clearActivePage(evt);
+    const selectorPageElementArray = this.element.querySelectorAll(this.options.selectorPageElement);
```
Naming nit - I think this is a node list instead of a selector?

asudoh commented on src/components/pagination-nav/pagination-nav.js:
```diff
@@ -0,0 +1,200 @@
+        this.activePageNumber += 1;
+      }
+    }
+    const selectorPageTarget = selectorPageElementArray[this.activePageNumber - 1];
```
Naming nit - I think this is a DOM element instead of a selector?

asudoh commented on src/components/pagination-nav/pagination-nav.js:
```diff
@@ -0,0 +1,200 @@
+   * Set Previous and Next button states
+   */
+  setPrevNextStates = () => {
+    const selectorPageElementArray = this.element.querySelectorAll(this.options.selectorPageElement);
```
Naming nit - I think this is a node list instead of a selector?

asudoh commented on src/components/pagination-nav/pagination-nav.js:
```diff
@@ -0,0 +1,200 @@
+  setPrevNextStates = () => {
+    const selectorPageElementArray = this.element.querySelectorAll(this.options.selectorPageElement);
+    const totalPages = selectorPageElementArray.length;
+    const selectorPageDirectionPrevious = this.element.querySelector('[data-page-direction="previous"]');
```
Naming nit - I think this is a DOM element instead of a selector?

asudoh commented on src/components/pagination-nav/pagination-nav.js:
```diff
@@ -0,0 +1,200 @@
+    const selectorPageElementArray = this.element.querySelectorAll(this.options.selectorPageElement);
+    const totalPages = selectorPageElementArray.length;
+    const selectorPageDirectionPrevious = this.element.querySelector('[data-page-direction="previous"]');
+    const selectorPageDirectionNext = this.element.querySelector('[data-page-direction="next"]');
```
Naming nit - I think this is a DOM element instead of a selector?

asudoh reviewed (commented):
Thank you for your quick update @Martskin! The latest code seems to be working very well. Just two series of comments - One is about avoiding caching (to cope with dynamic DOM update), another is naming nits.

Martskin committed:
Merge branch 'master' into 977-pagination-small

Martskin committed:
feat(pagination-nav): more js naming, cache tweaks per PR feedback

Martskin commented:
Got it! I really appreciate the thorough, clear direction! @asudoh
I made your last tweaks and pushed.

Martskin committed:
Merge branch 'master' into 977-pagination-small

asudoh commented on src/components/pagination-nav/pagination-nav.js:
```diff
@@ -0,0 +1,208 @@
+   */
+  clearActivePage = evt => {
+    const pageButtonNodeList = this.element.querySelectorAll(this.options.selectorPageButton);
+    const selectorPageSelect = this.element.querySelector(this.options.selectorPageSelect);
```
Naming nit - I think this is a DOM node instead of a selector?

asudoh commented on src/components/pagination-nav/pagination-nav.js:
```diff
@@ -0,0 +1,208 @@
+  setPrevNextStates = () => {
+    const pageElementNodeList = this.element.querySelectorAll(this.options.selectorPageElement);
+    const totalPages = pageElementNodeList.length;
+    const pageDirectionElementPrevious = this.element.querySelector(`[${this.options.attribPageDirection}="previous"]`);
```
Though this looks a neat technique, our codebase uses separate options for attributes/selectors. Doing the same here will be great.

asudoh reviewed (commented):
Close to completion, just a couple - Thanks @Martskin for another quick update!

Martskin committed:
Merge branch 'master' into 977-pagination-small

Martskin committed:
feat(pagination-nav): js naming and selector tweaks

Martskin committed:
feat(pagination-nav): js selector naming normalization

Martskin commented:
Thanks @asudoh ! 
I pushed another round of tweaks for review.

asudoh reviewed (approved):
LGTM 👍 - Thanks you for all the updates @Martskin!

Martskin commented:
Thanks! I'm happy to contribute!

Martskin committed:
Merge branch 'master' into 977-pagination-small

aagonzales commented:
Sorry, I should have taken a closer look at the designs when I 👍 this. The pattern in #977 is actually closest the original V9 design of this component. The V10 has some visual tweaks. This is really awesome to see though!

**V9 style fixes:**
- [ ] use `$brand-01` for the selected page indicator
- [ ] use `font-weight: 600` for the page numbers.
- [ ] use 16px of padding between page numbers

![image](https://user-images.githubusercontent.com/11670886/53034583-1f096380-3439-11e9-9579-c1dda211abe6.png)


**For V10:**
- [ ] use 4px x 16px underline for selected page (use color token $interactive-01). Line should be 12px from the bottom of the page number
- [ ] use 20x20 `caret` icon instead of `chevron`
- [ ] Add 32px padding between page numbers

![image](https://user-images.githubusercontent.com/11670886/53034588-23ce1780-3439-11e9-87d2-5d916a08edfe.png)

Martskin commented:
> Sorry, I should have taken a closer look at the designs when I 👍 this. The pattern in #977 is actually closest the original V9 design of this component. The V10 has some visual tweaks. This is really awesome to see though!

No worries @aagonzales ! These visual tweaks should be np. Thanks for the clear direction and feedback.

joshblack commented:
Thanks again for taking the time to contribute @Martskin! 🎉 We really appreciate it 🙏

tw15egan committed:
Merge branch 'master' into 977-pagination-small

Martskin committed:
Merge branch 'master' into 977-pagination-small

Martskin committed:
feat(pagination-nav): adds aria-disabled, aria-current. updates css

Martskin committed:
Merge branch 'master' into 977-pagination-small

Martskin commented:
Okay @asudoh @aagonzales . I updated the visual styles and pushed up my changes.

After testing this component in VoiceOver I changed all `disabled` attributes to `aria-disabled="true"` and added `aria-current="page"`. When using `disabled` the buttons/links aren't focusable. With `aria-disabled="true"` they are still able to receive focus and are read as "dimmed, current page" by VoiceOver.

https://deploy-preview-1812--the-carbon-components.netlify.com/?nav=pagination-nav

aagonzales commented:
Wow, its looking good! I love the animation on the selected line. Just two tweaks that I missed in my design spec. First can we make the selected page number use `font-weight: 600` / SemiBold. 

And then the originally designs had these big 48x48 backgrounds for the hover. Would it be possible to do something like that, instead of the small hovers?

![image](https://user-images.githubusercontent.com/11670886/53208787-bae3cc80-35fd-11e9-9fab-ef9b516214ec.png)


If not then I think instead of doing a background color on hover we should just change the text color on hover to `$interactive-01` the Blue 60 color.
![image](https://user-images.githubusercontent.com/11670886/53208816-dea71280-35fd-11e9-9824-c450773bfbe8.png)

Martskin committed:
Merge branch 'master' into 977-pagination-small

Martskin committed:
feat(pagination-nav): makes click target 48x48 on page buttons

Martskin commented:
> Wow, its looking good! I love the animation on the selected line. Just two tweaks that I missed in my design spec. First can we make the selected page number use `font-weight: 600` / SemiBold.
> 
> And then the originally designs had these big 48x48 backgrounds for the hover. Would it be possible to do something like that, instead of the small hovers?

Much better @aagonzales ! I like the bigger click target on the page buttons. Pushed for review:
https://deploy-preview-1812--the-carbon-components.netlify.com/?nav=pagination-nav

aagonzales reviewed (approved):
Awesome! It looks amazing!! Thanks for contributing, its so good to see 🎉 💯

Martskin commented:
Thanks! It's my pleasure.

jnm2377 commented:
This is awesome! Thanks for the contribution! 🎉 🎉
